src/java/org/apache/cassandra/journal/Journal.java

```diff
@@ -542,8 +542,11 @@ private void advanceSegment(ActiveSegment<K, V> oldSegment)
                 // if a segment is ready, take it now, otherwise wait for the allocator thread to construct it
                 if (availableSegment != null)
                 {
+                    ActiveSegment<K, V> nextSegment = availableSegment;
+                    availableSegment = null;
                     // success - change allocatingFrom and activeSegments (which must be kept in order) before leaving the critical section
-                    addNewActiveSegment(currentSegment = availableSegment);
+                    addNewActiveSegment(nextSegment);
```

Comment:
The update between `currentSegment` and the `Segments` can't happen simultaneously, unfortunately, so we have to make a decision regarding which happens first. Current order is intentional. The comment above explicitly says so (though variable names need an update). Code that is potentially affected by this needs to take this possibility into account. `Journal#getActiveSegment()` is an example of this - there, we check `currentSegment` first, and then `segments()` if need be. Similarly, `Journal#isFlushed()` must do the same thing.